realm/realm-library/src/main/java/io/realm/internal/Table.java

```diff
@@ -196,14 +198,35 @@ public long addColumnLink (RealmFieldType type, String name, Table table) {
     @Override
     public void removeColumn(long columnIndex) {
         nativeRemoveColumn(nativePtr, columnIndex);
+        // When PK does not exist this cannot happen. But if you remove a column with a smaller index
+        // than that of PK column, you need to rearrange PK column index.
+        if (hasPrimaryKey() && columnIndex <= getPrimaryKey()) {
+            invalidateCachedPrimaryKeyIndex();
+        }
     }
 
     /**
-     * Renames a column in the table.
+     * Renames a column in the table. This needs to be called manually before inserting data into the table.
+     *
+     * @param columnIndex the column index to be renamed.
+     * @param newName a new name replacing the old column name.
+     * @throws {@link IllegalArgumentException} if {@code newFieldName} is an empty string, or exceeds field name length limit.
      */
     @Override
     public void renameColumn(long columnIndex, String newName) {
         verifyColumnName(newName);
+        // a primary key renamed
+        if (getPrimaryKey() == columnIndex) {
+            if (TextUtils.isEmpty(newName)) {
+                throw new IllegalArgumentException("Primary key field cannot be renamed with an empty string.");
```

Comment:
I've modified `verifyColumnName()` itself to check an empty or long column name. It will affect
`addColumn()`, `addColumnLink()`, and `renameColumn()`.